File: DUUIWeb/src/lib/duui/pipeline.ts
import { v4 as uuidv4 } from 'uuid'
import { componentToJson, type DUUIComponent } from './component'

export interface DUUIPipeline {
	oid: string
	name: string
	description: string
	status: string
	tags: string[]
	created_at: number
	modified_at: number
	times_used: number
	last_used: number | undefined
	settings: Object
	user_id: string | null // if null -> Template
	components: DUUIComponent[]
	statistics: {
		status: AggreationResult
		errors: AggreationResult
		input: AggreationResult
		output: AggreationResult
		usage: AggreationResult
		size: AggreationResult
	}
}

/**
 * Create and return a default pipeline.
 *
 * @returns A default pipeline with no components.
 */
export const blankPipeline = () =>
	<DUUIPipeline>{
		oid: uuidv4(),
		name: 'New Pipeline',
		description: '',
		created_at: Date.now(),
		modified_at: Date.now(),
		times_used: 0,
		last_used: 0,
		settings: {},
		user_id: null,
		components: [],
		tags: [],
		status: 'Inactive',
		statistics: {
			status: [],
			errors: [],
			input: [],
			output: [],
			usage: [],
			size: []
		}
	}

/**
 * Convert a DUUIPipeline to a JSON object with some properties omitted.
 *
 * @param pipeline the pipeline to convert
 * @returns An object with relevant pipeline properties.
 */
export const pipelineToJson = (pipeline: DUUIPipeline) => {
	return {
		name: pipeline.name,
		description: pipeline.description,
		settings: pipeline.settings,
		components: pipeline.components.map((component) => componentToJson(component))
	}
}

/**
 * Retrieve a set of drivers that are used by components in the pipeline.
 *
 * @param pipeline The pipeline to get the drivers for
 * @returns A set of DUUIDrivers
 */
export const usedDrivers = (pipeline: DUUIPipeline) =>
	new Set(pipeline.components.map((c) => c.driver))

/**
 * Retrieve all pipeline tags as a string.
 *
 * @param pipeline The pipeline to get the tags for.
 * @returns A string containing all tags in the pipeline separated by spaces.
 */
export const getPipelineTagsAsString = (pipeline: DUUIPipeline) => {
	return pipeline.components.map((c) => c.tags.join(' ')).join(' ')
}

File: PythonClient/duui/objects/pipeline.py
from __future__ import annotations
import json

from duui.objects.component import DUUIComponent


class DUUIPipeline:
    def __init__(self, data: dict) -> None:
        self.__data = {}

        for key, value in data.items():
            if key not in [
                "oid",
                "name",
                "description",
                "status",
                "categories",
                "created_at",
                "modified_at",
                "times_used",
                "last_used",
                "settings",
                "components",
                "pipeline_id",
                "user_id",
            ]:
                continue

            self.__data[key] = value

    def __str__(self) -> str:
        return json.dumps(self.__data, indent=2, ensure_ascii=False)

    def __repr__(self) -> str:
        return f"DUUIPipeline(oid={self.oid}, modified_at={self.modified_at if self.modified_at > 0 else 'Never'})"

    @property
    def oid(self) -> str:
        return self.__data.get("oid", "")

    @property
    def name(self) -> str:
        return self.__data.get("name", "")

    @property
    def description(self) -> str:
        return self.__data.get("description", "")

    @property
    def categories(self) -> str:
        return self.__data.get("categories", [])

    @property
    def status(self) -> str:
        return self.__data.get("status", "")

    @property
    def created_at(self) -> int:
        return self.__data.get("created_at", -1)

    @property
    def modified_at(self) -> int:
        return self.__data.get("modified_at", -1)

    @property
    def times_used(self) -> int:
        return self.__data.get("times_used", 0)

    @property
    def last_used(self) -> int:
        return self.__data.get("last_used", -1)

    @property
    def settings(self) -> dict:
        return self.__data.get("settings", {})

    @property
    def components(self) -> list[DUUIComponent]:
        return self.__data.get("components", [])

    @property
    def pipeline_id(self) -> str:
        return self.__data.get("pipeline_id", "")

    @property
    def user_id(self) -> str:
        return self.__data.get("user_id", "")
